tsnake/initialize.py

```diff
@@ -0,0 +1,543 @@
+"""
+Module for reading in image+mask, then initializing the T-snakes accordingly.
+
+NOTE: we might need to add some kind of padding to the mask-rectangles?
+"""
+import os
+
+import cv2
+import numpy as np
+import matplotlib.pyplot as plt
+from scipy.ndimage import gaussian_filter
+from scipy.spatial.distance import cdist
+
+from .snake import Node, TSnake
+
+
+# =====================================================
+# FOR LOADING IMAGE AND MASK FILES
+# =====================================================
+
+def load_grayscale_image(path):
+    """
+    Args:
+    ==============
+    (str) path:
+    * Path to *raw* image file.
+    ==============
+    
+    Returns:
+    2D np.array representing *grayscale* image.
+    """
+    path = os.path.abspath(path)
+    image = cv2.imread(path, 0)
+    assert image is not None, 'failed to load image at path=%s' % path
+    
+    # sanity checks
+    assert image.max() <= 255
+    assert image.min() >= 0
+    assert len(image.shape) == 2
+    
+    return image
+
+
+def load_mask(path):
+    """
+    Args:
+    ==============
+    (str) path:
+    * Path to mask file.
+    ==============
+    
+    Returns:
+    2D binary np.array representing the mask.
+      - mask[i, j] == 1    --> means that pixel (i, j) is masked
+    """
+    path = os.path.abspath(path)
+    image = cv2.imread(path)
+    assert image is not None, 'failed to load mask at path=%s' % path
+    
+    # Some asserts to make sure the input image is as expected
+    # Mask should be binary 0/255
+    assert set(np.unique(image)) == {0, 255}, set(np.unique(image))
+    # Mask should be 3D
+    assert len(image.shape) == 3
+    assert image.shape[2] == 3
+    
+    # throw away useless 3rd dimension
+    image = image[:, :, 0]
+    # replace 255 -> 1
+    image = np.where(image == 0, 0, 1)
+    
+    return image
+
+# =====================================================
+# FOR EXTRACTING DISTINCT RECTANGULAR MASKED REGIONS
+# =====================================================
+
+class _Region(object):
+    """
+    A rectangular mask region.
+    
+    This class is used as an intermediate for building MaskedRegion objects. 
+    """
+        
+    def __init__(self):
+        self.coords = set()
+
+        self.leftmost = np.inf
+        self.rightmost = -np.inf
+        self.uppermost = -np.inf
+        self.bottommost = np.inf
+
+    def _is_member(self, r, c):
+        for r1 in range(r-1, r+2):
+            for c1 in range(c-1, c+2):
+                if (r1, c1) in self.coords:
+                    return True
+
+        return False
+
+    def _add_coordinate(self, r, c):
+        self.coords.add((r, c))
+
+        self.leftmost = min(self.leftmost, c)
+        self.rightmost = max(self.rightmost, c)
+        self.uppermost = max(self.uppermost, r)
```

Comment:
How do you view the image coordinate system? I assumed that (0,0) was the upper left corner of the image and X-coordinate represents the rows and Y-coordinate represents the columns.

You take the max for the uppermost point, so is the uppermost the top of the image (in my representation, the lower the x-coordinate, the higher up the image)?

Reply:
Yeah so I accidentally named the variables in the `_Region` class a bit confusingly (should have been more careful about that tbh), but the `Region` code still works correctly in extracting the right part of the image. 

Since we shouldn't need to edit `_Region` any more, I figured we can just leave it as is